src/lib.rs

```diff
@@ -727,15 +727,216 @@ impl CertificateParams {
+		let subject_alt_names = Self::convert_x509_subject_alternative_name(&x509)?;
+		let key_usages = Self::convert_x509_key_usages(&x509)?;
+		let extended_key_usages = Self::convert_x509_extended_key_usages(&x509)?;
+		let name_constraints = Self::convert_x509_name_constraints(&x509)?;
+
 		Ok(
 			CertificateParams {
 				alg,
+				is_ca,
+				subject_alt_names,
+				key_usages,
+				extended_key_usages,
+				name_constraints,
 				distinguished_name : dn,
 				key_pair : Some(key_pair),
+				serial_number: Some(serial),
+				not_before : validity.not_before.to_datetime(),
+				not_after : validity.not_after.to_datetime(),
 				.. Default::default()
 			}
 		)
 	}
+	#[cfg(feature = "x509-parser")]
+	fn convert_x509_is_ca(x509 :&x509_parser::certificate::X509Certificate<'_>) -> Result<IsCa, RcgenError> {
+		use x509_parser::extensions::BasicConstraints as B;
+
+		let basic_constraints = x509.basic_constraints()
+			.or(Err(RcgenError::CouldNotParseCertificate))?.map(|ext| ext.value);
+
+		let is_ca = match basic_constraints {
+			Some(B { ca: true, path_len_constraint: Some(n) }) if *n <= u8::MAX as u32 => IsCa::Ca(BasicConstraints::Constrained(*n as u8)),
+			Some(B { ca: true, path_len_constraint: Some(_) }) => return Err(RcgenError::CouldNotParseCertificate),
+			Some(B { ca: true, path_len_constraint: None }) => IsCa::Ca(BasicConstraints::Unconstrained),
+			Some(B { ca: false, .. }) => IsCa::ExplicitNoCa,
+			None => IsCa::NoCa,
+		};
+
+		Ok(is_ca)
+	}
+	#[cfg(feature = "x509-parser")]
+	fn convert_x509_serial(x509 :&x509_parser::certificate::X509Certificate<'_>) -> Result<u64, RcgenError> {
+		const SIZE: usize = std::mem::size_of::<u64>();
+
+		let raw = x509.raw_serial();
+		let len = raw.len();
+
+		if len > SIZE {
+			return Err(RcgenError::CouldNotParseCertificate);
+		}
+
+		let mut bytes = [0u8; SIZE];
+		bytes[(SIZE - len)..].copy_from_slice(raw);
+		Ok(u64::from_be_bytes(bytes))
+	}
+	#[cfg(feature = "x509-parser")]
+	fn convert_x509_subject_alternative_name(x509 :&x509_parser::certificate::X509Certificate<'_>) -> Result<Vec<SanType>, RcgenError> {
+		let sans = x509.subject_alternative_name()
+			.or(Err(RcgenError::CouldNotParseCertificate))?
+			.map(|ext| &ext.value.general_names);
+
+		if let Some(sans) = sans {
+			let mut subject_alt_names = Vec::with_capacity(sans.len());
+			for san in sans {
+				use x509_parser::extensions::GeneralName;
+				let name = match san {
+					GeneralName::RFC822Name(s) => SanType::Rfc822Name(s.to_string()),
+					GeneralName::DNSName(s) => SanType::DnsName(s.to_string()),
+					GeneralName::URI(s) => SanType::URI(s.to_string()),
+					GeneralName::IPAddress(bytes) if bytes.len() == 4 => {
+						let bytes: [u8; 4] = (*bytes).try_into().unwrap();
+						SanType::IpAddress(IpAddr::from(bytes))
+					}
+					GeneralName::IPAddress(bytes) if bytes.len() == 16 => {
+						let bytes: [u8; 16] = (*bytes).try_into().unwrap();
+						SanType::IpAddress(IpAddr::from(bytes))
+					}
+					_ => continue,
```

Comment:
Why wouldn't we want the compiler to flag that an implementation is needed here if a new entry for `GeneralName` is added in the future?